src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokenDocumentationProvider.java

```diff
@@ -0,0 +1,118 @@
+
+    @Override
+    @Nullable
+    @Nls
+    public String getQuickNavigateInfo(PsiElement targetElement,
+                                       PsiElement sourceElement) {
+        String documentation = getDocumentation(sourceElement, targetElement);
+        return documentation != null ? documentation : super.getQuickNavigateInfo(targetElement, sourceElement);
+    }
+
+    /**
+     * Provides external access to documentation for other parts of LSP4IJ, e.g.,
+     * {@link com.redhat.devtools.lsp4ij.usages.LSPUsageElementDescriptionProvider}.
+     *
+     * @param element the element for which documentation is requested
+     * @return documentation for the element, or null if none could be found
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public static String getDocumentation(@Nullable PsiElement element) {
+        return getDocumentation(element, element);
+    }
+
+    @Nullable
+    private static String getDocumentation(@Nullable PsiElement sourceElement,
+                                           @Nullable PsiElement targetElement) {
+        LSPSemanticTokensFileViewProvider semanticTokensFileViewProvider = LSPSemanticTokensFileViewProvider.getInstance(sourceElement);
+        if (semanticTokensFileViewProvider != null) {
+            int offset = sourceElement.getTextOffset();
+            ThreeState isIdentifier = semanticTokensFileViewProvider.isIdentifier(offset);
+            if ((isIdentifier == ThreeState.YES) ||
+                // If the token is definitely not an identifier and it's not whitespace, give the benefit of the doubt
+                ((isIdentifier == ThreeState.NO) && !semanticTokensFileViewProvider.isWhitespace(offset))) {
+                // Try to get the actual documentation for the element via LSP
+                try {
+                    PsiFile file = sourceElement.getContainingFile();
+                    List<LSPDocumentationTarget> lspDocumentationTargets = LSP_DOCUMENTATION_TARGET_PROVIDER
+                            .documentationTargets(file, offset)
+                            .stream()
+                            .filter((Predicate<DocumentationTarget>) dt -> dt instanceof LSPDocumentationTarget)
+                            .map((Function<DocumentationTarget, LSPDocumentationTarget>) dt -> (LSPDocumentationTarget) dt)
+                            .toList();
+                    if (!ContainerUtil.isEmpty(lspDocumentationTargets)) {
+                        String documentation = StringUtil.join(lspDocumentationTargets, LSPDocumentationTarget::getHtml, "\n").trim();
+                        if (StringUtil.isNotEmpty(documentation)) {
+                            return documentation;
+                        }
+                    }
+                } catch (ProcessCanceledException pce) {
+                    // Larger documentation can time out, so just degrade gracefully to fallback documentation
```

Comment:
I saw timeouts for Java imports via JDTLS where the entire class header doc was being returned and converted from Markdown to HTML. If that happens, we'll just degrade to the client-side-only fallback doc approach below.